File: flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamJoinOperator.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.streaming.api.scala

import java.util.concurrent.TimeUnit

import org.apache.flink.api.common.ExecutionConfig
import org.apache.flink.api.common.functions.JoinFunction
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.api.common.typeutils.TypeSerializer
import org.apache.flink.api.java.functions.KeySelector
import org.apache.flink.api.java.operators.Keys
import org.apache.flink.api.scala.typeutils.{CaseClassSerializer, CaseClassTypeInfo}
import org.apache.flink.streaming.api.datastream.temporal.TemporalWindow
import org.apache.flink.streaming.api.datastream.{DataStream => JavaStream, SingleOutputStreamOperator}
import org.apache.flink.streaming.api.functions.co.JoinWindowFunction
import org.apache.flink.streaming.api.operators.co.CoStreamWindow
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment.clean
import org.apache.flink.streaming.util.keys.KeySelectorUtil

import scala.Array.canBuildFrom
import scala.reflect.ClassTag

class StreamJoinOperator[I1, I2](i1: JavaStream[I1], i2: JavaStream[I2]) extends 
TemporalOperator[I1, I2, StreamJoinOperator.JoinWindow[I1, I2]](i1, i2) {

  override def createNextWindowOperator() = {
    new StreamJoinOperator.JoinWindow[I1, I2](this)
  }
}

object StreamJoinOperator {

  class JoinWindow[I1, I2](private[flink] val op: StreamJoinOperator[I1, I2]) extends
  TemporalWindow[JoinWindow[I1, I2]] {

    private[flink] val type1 = op.input1.getType()

    /**
     * Continues a temporal Join transformation by defining
     * the fields in the first stream to be used as keys for the join.
     * The resulting incomplete join can be completed by JoinPredicate.equalTo()
     * to define the second key.
     */
    def where(fields: Int*) = {
      new JoinPredicate[I1, I2](op, KeySelectorUtil.getSelectorForKeys(
        new Keys.ExpressionKeys(fields.toArray, type1),
        type1,
        op.input1.getExecutionEnvironment.getConfig))
    }

    /**
     * Continues a temporal Join transformation by defining
     * the fields in the first stream to be used as keys for the join.
     * The resulting incomplete join can be completed by JoinPredicate.equalTo()
     * to define the second key.
     */
    def where(firstField: String, otherFields: String*) =
      new JoinPredicate[I1, I2](op, KeySelectorUtil.getSelectorForKeys(
        new Keys.ExpressionKeys(firstField +: otherFields.toArray, type1),
        type1,
        op.input1.getExecutionEnvironment.getConfig))

    /**
     * Continues a temporal Join transformation by defining
     * the keyselector function that will be used to extract keys from the first stream
     * for the join.
     * The resulting incomplete join can be completed by JoinPredicate.equalTo()
     * to define the second key.
     */
    def where[K: TypeInformation](fun: (I1) => K) = {
      val keyType = implicitly[TypeInformation[K]]
      val keyExtractor = new KeySelector[I1, K] {
        val cleanFun = op.input1.clean(fun)
        def getKey(in: I1) = cleanFun(in)
      }
      new JoinPredicate[I1, I2](op, keyExtractor)
    }

    override def every(length: Long, timeUnit: TimeUnit): JoinWindow[I1, I2] = {
      every(timeUnit.toMillis(length))
    }

    override def every(length: Long): JoinWindow[I1, I2] = {
      op.slideInterval = length
      this
    }

  }

  class JoinPredicate[I1, I2](private[flink] val op: StreamJoinOperator[I1, I2],
    private[flink] val keys1: KeySelector[I1, _]) {
    private[flink] var keys2: KeySelector[I2, _] = null
    private[flink] val type2 = op.input2.getType()

    /**
     * Creates a temporal join transformation by defining the second join key.
     * The returned transformation wrapes each joined element pair in a tuple2:
     * (first, second)
     * To define a custom wrapping, use JoinedStream.apply(...)
     */
    def equalTo(fields: Int*): JoinedStream[I1, I2] = {
      finish(KeySelectorUtil.getSelectorForKeys(
        new Keys.ExpressionKeys(fields.toArray, type2),
        type2,
        op.input1.getExecutionEnvironment.getConfig))
    }

    /**
     * Creates a temporal join transformation by defining the second join key.
     * The returned transformation wrapes each joined element pair in a tuple2:
     * (first, second)
     * To define a custom wrapping, use JoinedStream.apply(...)
     */
    def equalTo(firstField: String, otherFields: String*): JoinedStream[I1, I2] =
      finish(KeySelectorUtil.getSelectorForKeys(
        new Keys.ExpressionKeys(firstField +: otherFields.toArray, type2),
        type2,
        op.input1.getExecutionEnvironment.getConfig))

    /**
     * Creates a temporal join transformation by defining the second join key.
     * The returned transformation wrapes each joined element pair in a tuple2:
     * (first, second)
     * To define a custom wrapping, use JoinedStream.apply(...)
     */
    def equalTo[K: TypeInformation](fun: (I2) => K): JoinedStream[I1, I2] = {
      val keyType = implicitly[TypeInformation[K]]
      val keyExtractor = new KeySelector[I2, K] {
        val cleanFun = op.input1.clean(fun)
        def getKey(in: I2) = cleanFun(in)
      }
      finish(keyExtractor)
    }

    private def finish(keys2: KeySelector[I2, _]): JoinedStream[I1, I2] = {
      this.keys2 = keys2
      new JoinedStream[I1, I2](this, createJoinOperator())
    }

    private def createJoinOperator(): JavaStream[(I1, I2)] = {

      val returnType = new CaseClassTypeInfo[(I1, I2)](
        classOf[(I1, I2)],
        Array(op.input1.getType, op.input2.getType),
        Seq(op.input1.getType, op.input2.getType),
        Array("_1", "_2")) {

        override def createSerializer(
            executionConfig: ExecutionConfig): TypeSerializer[(I1, I2)] = {
          val fieldSerializers: Array[TypeSerializer[_]] = new Array[TypeSerializer[_]](getArity)
          for (i <- 0 until getArity) {
            fieldSerializers(i) = types(i).createSerializer(executionConfig)
          }

          new CaseClassSerializer[(I1, I2)](classOf[(I1, I2)], fieldSerializers) {
            override def createInstance(fields: Array[AnyRef]) = {
              (fields(0).asInstanceOf[I1], fields(1).asInstanceOf[I2])
            }
          }
        }
      }

      op.input1.groupBy(keys1).connect(op.input2.groupBy(keys2))
        .addGeneralWindowCombine(getJoinWindowFunction(this, (_, _)),
          returnType, op.windowSize, op.slideInterval, op.timeStamp1, op.timeStamp2)
    }
  }

  class JoinedStream[I1, I2](jp: JoinPredicate[I1, I2], javaStream: JavaStream[(I1, I2)]) extends 
  DataStream[(I1, I2)](javaStream) {

    private val op = jp.op

    /**
     * Sets a wrapper for the joined elements. For each joined pair, the result of the
     * udf call will be emitted.
     */
    def apply[R: TypeInformation: ClassTag](fun: (I1, I2) => R): DataStream[R] = {

      val operator = new CoStreamWindow[I1, I2, R](
        clean(getJoinWindowFunction(jp, fun)), op.windowSize, op.slideInterval, op.timeStamp1,
        op.timeStamp2)

      javaStream.getExecutionEnvironment().getStreamGraph().setOperator(javaStream.getId(),
        operator)

      val js = javaStream.asInstanceOf[SingleOutputStreamOperator[R,_]]
      js.returns(implicitly[TypeInformation[R]]).asInstanceOf[SingleOutputStreamOperator[R,_]]
    }
  }

  private[flink] def getJoinWindowFunction[I1, I2, R](jp: JoinPredicate[I1, I2],
    joinFunction: (I1, I2) => R) = {
    require(joinFunction != null, "Join function must not be null.")

    val joinFun = new JoinFunction[I1, I2, R] {

      val cleanFun = jp.op.input1.clean(joinFunction)

      override def join(first: I1, second: I2): R = {
        cleanFun(first, second)
      }
    }

    new JoinWindowFunction[I1, I2, R](jp.keys1, jp.keys2, joinFun)
  }

}

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/temporal/StreamJoinOperator.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.streaming.api.datastream.temporal;

import java.util.concurrent.TimeUnit;

import org.apache.flink.api.common.functions.JoinFunction;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.api.java.operators.Keys;
import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.typeutils.TupleTypeInfo;
import org.apache.flink.api.java.typeutils.TypeExtractor;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.functions.co.JoinWindowFunction;
import org.apache.flink.streaming.api.operators.co.CoStreamWindow;
import org.apache.flink.streaming.util.keys.KeySelectorUtil;

public class StreamJoinOperator<I1, I2> extends
		TemporalOperator<I1, I2, StreamJoinOperator.JoinWindow<I1, I2>> {

	public StreamJoinOperator(DataStream<I1> input1, DataStream<I2> input2) {
		super(input1, input2);
	}

	@Override
	protected JoinWindow<I1, I2> createNextWindowOperator() {
		return new JoinWindow<I1, I2>(this);
	}

	public static class JoinWindow<I1, I2> implements TemporalWindow<JoinWindow<I1, I2>> {

		private StreamJoinOperator<I1, I2> op;
		private TypeInformation<I1> type1;

		private JoinWindow(StreamJoinOperator<I1, I2> operator) {
			this.op = operator;
			this.type1 = op.input1.getType();
		}

		/**
		 * Continues a temporal Join transformation. <br/>
		 * Defines the {@link Tuple} fields of the first join {@link DataStream}
		 * that should be used as join keys.<br/>
		 * <b>Note: Fields can only be selected as join keys on Tuple
		 * DataStreams.</b><br/>
		 * 
		 * @param fields
		 *            The indexes of the other Tuple fields of the first join
		 *            DataStreams that should be used as keys.
		 * @return An incomplete Join transformation. Call
		 *         {@link JoinPredicate#equalTo} to continue the Join.
		 */
		public JoinPredicate<I1, I2> where(int... fields) {
			return new JoinPredicate<I1, I2>(op, KeySelectorUtil.getSelectorForKeys(
					new Keys.ExpressionKeys<I1>(fields, type1), type1, op.input1.getExecutionEnvironment().getConfig()));
		}

		/**
		 * Continues a temporal join transformation. <br/>
		 * Defines the fields of the first join {@link DataStream} that should
		 * be used as grouping keys. Fields are the names of member fields of
		 * the underlying type of the data stream.
		 * 
		 * @param fields
		 *            The fields of the first join DataStream that should be
		 *            used as keys.
		 * @return An incomplete Join transformation. Call
		 *         {@link JoinPredicate#equalTo} to continue the Join.
		 */
		public JoinPredicate<I1, I2> where(String... fields) {
			return new JoinPredicate<I1, I2>(op, KeySelectorUtil.getSelectorForKeys(
					new Keys.ExpressionKeys<I1>(fields, type1), type1, op.input1.getExecutionEnvironment().getConfig()));
		}

		/**
		 * Continues a temporal Join transformation and defines a
		 * {@link KeySelector} function for the first join {@link DataStream}
		 * .</br> The KeySelector function is called for each element of the
		 * first DataStream and extracts a single key value on which the
		 * DataStream is joined. </br>
		 * 
		 * @param keySelector
		 *            The KeySelector function which extracts the key values
		 *            from the DataStream on which it is joined.
		 * @return An incomplete Join transformation. Call
		 *         {@link JoinPredicate#equalTo} to continue the Join.
		 */
		public <K> JoinPredicate<I1, I2> where(KeySelector<I1, K> keySelector) {
			return new JoinPredicate<I1, I2>(op, keySelector);
		}

		@Override
		public JoinWindow<I1, I2> every(long length, TimeUnit timeUnit) {
			return every(timeUnit.toMillis(length));
		}

		@Override
		public JoinWindow<I1, I2> every(long length) {
			op.slideInterval = length;
			return this;
		}

		// ----------------------------------------------------------------------------------------

	}

	/**
	 * Intermediate step of a temporal Join transformation. <br/>
	 * To continue the Join transformation, select the join key of the second
	 * input {@link DataStream} by calling {@link JoinPredicate#equalTo}
	 * 
	 */
	public static class JoinPredicate<I1, I2> {

		private StreamJoinOperator<I1, I2> op;
		private KeySelector<I1, ?> keys1;
		private KeySelector<I2, ?> keys2;
		private TypeInformation<I2> type2;

		private JoinPredicate(StreamJoinOperator<I1, I2> operator, KeySelector<I1, ?> keys1) {
			this.op = operator;
			this.keys1 = keys1;
			this.type2 = op.input2.getType();
		}

		/**
		 * Creates a temporal Join transformation and defines the {@link Tuple}
		 * fields of the second join {@link DataStream} that should be used as
		 * join keys.<br/>
		 * </p> The resulting operator wraps each pair of joining elements in a
		 * Tuple2<I1,I2>(first, second). To use a different wrapping function
		 * use {@link JoinedStream#with(JoinFunction)}
		 * 
		 * @param fields
		 *            The indexes of the Tuple fields of the second join
		 *            DataStream that should be used as keys.
		 * @return A streaming join operator. Call {@link JoinedStream#with} to
		 *         apply a custom wrapping
		 */
		public JoinedStream<I1, I2> equalTo(int... fields) {
			keys2 = KeySelectorUtil.getSelectorForKeys(new Keys.ExpressionKeys<I2>(fields, type2),
					type2, op.input1.getExecutionEnvironment().getConfig());
			return createJoinOperator();
		}

		/**
		 * Creates a temporal Join transformation and defines the fields of the
		 * second join {@link DataStream} that should be used as join keys. </p>
		 * The resulting operator wraps each pair of joining elements in a
		 * Tuple2<I1,I2>(first, second). To use a different wrapping function
		 * use {@link JoinedStream#with(JoinFunction)}
		 * 
		 * @param fields
		 *            The fields of the second join DataStream that should be
		 *            used as keys.
		 * @return A streaming join operator. Call {@link JoinedStream#with} to
		 *         apply a custom wrapping
		 */
		public JoinedStream<I1, I2> equalTo(String... fields) {
			this.keys2 = KeySelectorUtil.getSelectorForKeys(new Keys.ExpressionKeys<I2>(fields,
					type2), type2, op.input1.getExecutionEnvironment().getConfig());
			return createJoinOperator();
		}

		/**
		 * Creates a temporal Join transformation and defines a
		 * {@link KeySelector} function for the second join {@link DataStream}
		 * .</br> The KeySelector function is called for each element of the
		 * second DataStream and extracts a single key value on which the
		 * DataStream is joined. </p> The resulting operator wraps each pair of
		 * joining elements in a Tuple2<I1,I2>(first, second). To use a
		 * different wrapping function use
		 * {@link JoinedStream#with(JoinFunction)}
		 * 
		 * 
		 * @param keySelector
		 *            The KeySelector function which extracts the key values
		 *            from the second DataStream on which it is joined.
		 * @return A streaming join operator. Call {@link JoinedStream#with} to
		 *         apply a custom wrapping
		 */
		public <K> JoinedStream<I1, I2> equalTo(KeySelector<I2, K> keySelector) {
			this.keys2 = keySelector;
			return createJoinOperator();
		}

		private JoinedStream<I1, I2> createJoinOperator() {

			JoinFunction<I1, I2, Tuple2<I1, I2>> joinFunction = new DefaultJoinFunction<I1, I2>();

			JoinWindowFunction<I1, I2, Tuple2<I1, I2>> joinWindowFunction = getJoinWindowFunction(
					joinFunction, this);

			TypeInformation<Tuple2<I1, I2>> outType = new TupleTypeInfo<Tuple2<I1, I2>>(
					op.input1.getType(), op.input2.getType());

			return new JoinedStream<I1, I2>(this, op.input1
					.groupBy(keys1)
					.connect(op.input2.groupBy(keys2))
					.addGeneralWindowCombine(joinWindowFunction, outType, op.windowSize,
							op.slideInterval, op.timeStamp1, op.timeStamp2));
		}
	}

	public static class JoinedStream<I1, I2> extends
			SingleOutputStreamOperator<Tuple2<I1, I2>, JoinedStream<I1, I2>> {
		private final JoinPredicate<I1, I2> predicate;

		private JoinedStream(JoinPredicate<I1, I2> predicate, DataStream<Tuple2<I1, I2>> ds) {
			super(ds);
			this.predicate = predicate;
		}

		/**
		 * Completes a stream join. </p> The resulting operator wraps each pair
		 * of joining elements using the user defined {@link JoinFunction}
		 * 
		 * @return The joined data stream.
		 */
		@SuppressWarnings("unchecked")
		public <OUT> SingleOutputStreamOperator<OUT, ?> with(JoinFunction<I1, I2, OUT> joinFunction) {

			TypeInformation<OUT> outType = TypeExtractor.getJoinReturnTypes(joinFunction,
					predicate.op.input1.getType(), predicate.op.input2.getType());

			CoStreamWindow<I1, I2, OUT> operator = new CoStreamWindow<I1, I2, OUT>(
					getJoinWindowFunction(joinFunction, predicate), predicate.op.windowSize,
					predicate.op.slideInterval, predicate.op.timeStamp1, predicate.op.timeStamp2);

			streamGraph.setOperator(id, operator);

			return ((SingleOutputStreamOperator<OUT, ?>) this).returns(outType);
		}
	}

	public static final class DefaultJoinFunction<I1, I2> implements
			JoinFunction<I1, I2, Tuple2<I1, I2>> {

		private static final long serialVersionUID = 1L;
		private final Tuple2<I1, I2> outTuple = new Tuple2<I1, I2>();

		@Override
		public Tuple2<I1, I2> join(I1 first, I2 second) throws Exception {
			outTuple.f0 = first;
			outTuple.f1 = second;
			return outTuple;
		}
	}

	private static <I1, I2, OUT> JoinWindowFunction<I1, I2, OUT> getJoinWindowFunction(
			JoinFunction<I1, I2, OUT> joinFunction, JoinPredicate<I1, I2> predicate) {
		return new JoinWindowFunction<I1, I2, OUT>(predicate.keys1, predicate.keys2, joinFunction);
	}
}
